unhide uitextinput when focused

chunhtai opened the pull request:
To make the UITextInput work properly, its accessibilityElementsHidden must return NO.

This pr makes sure to toggle it on or off when it is focused. However, I am not sure if there will be any side effect when doing this. Will write test once we confirm this is what we want to do.

Fixes https://github.com/flutter/flutter/issues/71885

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.
- [ ] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1116,7 +1118,7 @@ - (BOOL)accessibilityElementsHidden {
   // mimic the semantics tree from Flutter. We want the text field to be represented as a
   // `TextInputSemanticsObject` in that `SemanticsObject` tree rather than in this
   // `FlutterTextInputView` bridge which doesn't appear above a text field from the Flutter side.
```
We should update this comment.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -424,6 +424,7 @@ @interface FlutterTextInputView ()
 @property(nonatomic, readonly) CATransform3D editableTransform;
 @property(nonatomic, assign) CGRect markedRect;
 @property(nonatomic) BOOL isVisibleToAutofill;
+@property(nonatomic) BOOL hidden;
```
This should be called something like `accessibilityHidden`.

gaaclarke reviewed (approved):
Nice detective work.  I can't think of why this wouldn't be an acceptable fix.  My biggest concern would be that we could swipe to have voiceover announce a widget that should be hidden after `showTextInput` is called but before `hideTextInput`.  I'd make sure that couldn't happen with manual testing.

chunhtai commented:
@gaaclarke thanks for reminding, this PR doesn't work, i am able to swipe to focus the FlutterTextInputView.

I couldn't find a way to be able to make voiceover respond to UITextInput event while keeping the FlutterTextInputView hidden from accessibility focus.

At this moment, I am thinking whether there is a way to point the TextInputSemanticsObject to the FlutterTextInputView

chinmaygarde commented:
@chunhtai: Based on your last comment, it looks like this PR is not quite ready to land. Is that accurate? Can we close it or mark it WIP?

chunhtai commented:
I have revised this pr. Test will be added once we confirm we want to proceed with this fix

There are three fail-safes I added to the `fluttertextinputview`

1. Add back the `fluttertextinputview` hider, This prevent `fluttertextinputview` from receiving swipping-to-focus
2. Adds delay enable logic to `fluttertextinputview`. It is a known [issue](https://github.com/flutter/flutter/issues/53065) that the `fluttertextinputview` may receive focus when it is activated during screen changes. This delay makes sure the voiceover cannot find the `fluttertextinputview` while it try to find the new focus due to screen changes.
3. `fluttertextinputview` will direct the focus back to the current active textfield if it ever receives focus. I added this fail-safe as the last resort. I have not able to focus the `fluttertextinputview` after adding the above two fail-safes, but i assumed it is still possible. If such unfortunate event happens, this will make sure the focus is not stuck at the `fluttertextinputview`.

chunhtai commented:
@gaaclarke @xster Can you take another look whether this is what we want to do? I will add more tests once we confirm that.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1207,11 +1259,33 @@ - (void)updateMarkedRect:(NSDictionary*)dictionary {
+  // with a semantics update sent to the engine. The voiceover will focus
+  // the newly attached active view while performing accessibility update.
+  // This results in accessibility focus stuck at the FlutterTextInputView.
+  [NSTimer scheduledTimerWithTimeInterval:_kUITextInputAccessibilityEnablingDelaySeconds
```
This causes the plugin to live beyond the point it would be dealloc'd previously.  If you do this you should maintain a reference to the timer and invalidate it in dealloc.

gaaclarke replied:
The proper constant name in objective-c doesn't have a `_` at the beginning.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -533,6 +541,14 @@ - (void)dealloc {
   [super dealloc];
 }
 
++ (UIAccessibilityElement*)backingTextInputAccessibilityObject {
+  return _backingTextInputAccessibilityObject;
+}
+
++ (void)setBackingTextInputAccessibilityObject:(UIAccessibilityElement*)element {
+  _backingTextInputAccessibilityObject = element;
+}
```
There is no need to make these methods, this should just be a property.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1106,16 +1122,49 @@ - (void)deleteBackward {
+ * This prevents the `FlutterTextInputView` from receiving the focus
+ * due to swipping gesture.
+ *
+ * There are other cases the `FlutterTextInputView` may receive
+ * focus. One example is during screen changes, the accessibility
+ * tree will undergo a dramatic structural update. The Voiceover may
+ * decide to focus the `FlutterTextInputView` that is not involved
+ * in the structural update instead. If that happens, the
+ * `FlutterTextInputView` will make a best effort to direct the
+ * focus back to the `SemanticsObject`.
```
Couldn't this result in duplicate Voiceover announcements?  One for the FlutterTextInputView and one SemanticsObject?

chunhtai replied:
that is true, that is why i put the first two fail safe and hope it can catch all of the corner cases.

If this text view does receive the focus, duplicate announcement is still much better than focus just stuck at this textview.

I have tried a number of different hack such as changing the trait and label of this text view to prevent voiceover from pronouncing any word when the text view is focused, but none of them work. If you have a textview without any trait or label, it will pronounce some random phrase such as "blue sky" when it is focused.

gaaclarke reviewed (commented):
Jeez, this is hard to review.  I understand it's a hack.  Why did you choose 0.5s?  If the problem we are trying to fix is the missing of the announcement of typed characters it seems like 0.5s could be ample time to miss a character.

chunhtai commented:
@gaaclarke Hi 0.5 seems to be the lowest possible timeout; otherwise, the may still focus the textfield. I am not sure if this is device specific, but if the other low end iOS device can still focus textfield with 0.5 delay, they will be caught in fail safe 3.

chunhtai commented:
I will do some more clean up, but i think this is the best we can do for now.

chinmaygarde commented:
@gaaclarke Does your previous approval need to be revised? If so, can you review again please?

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1128,10 +1170,13 @@ @interface FlutterTextInputPlugin ()
 
 @implementation FlutterTextInputPlugin
 
+NSTimer* _enableFlutterTextInputViewAccessibilityTimer;
+
```
This is a global variable as written, you need to put this in braces.

chunhtai replied:
oops nice catch. will update

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1150,11 +1196,18 @@ - (instancetype)init {
+- (void)removeEnableFlutterTextInputViewAccessibilityTimer {
+  if (_enableFlutterTextInputViewAccessibilityTimer) {
+    [_enableFlutterTextInputViewAccessibilityTimer invalidate];
+    _enableFlutterTextInputViewAccessibilityTimer = nil;
```
This is a leak, sorry i didn't see it earlier.  You need to retain this below, and release this here.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1150,11 +1196,19 @@ - (instancetype)init {
 - (void)dealloc {
   [self hideTextInput];
   [_reusableInputView release];
+  [_inputHider release];
```
there needs to be ` [_enableFlutterTextInputViewAccessibilityTimer release]` as well =)

chunhtai replied:
that will be trigger by hideTextInput, my thinking is that the timer is created in showtextinput, so it should be removed in hidetextinput

gaaclarke replied:
Yea, I think in practice you are right since the timer will retain the FlutterTextInputPlugin until the timer fires.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1207,11 +1261,38 @@ - (void)updateMarkedRect:(NSDictionary*)dictionary {
+  if (_activeView.isFirstResponder) {
+    _activeView.accessibilityEnabled = YES;
+  }
+  _enableFlutterTextInputViewAccessibilityTimer = nil;
```
there needs to be a release here... maybe it would just be easier if you used a property for this.

chunhtai replied:
thanks for reminding.

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:
```diff
@@ -1128,9 +1170,12 @@ @interface FlutterTextInputPlugin ()
 @property(nonatomic, readonly)
     NSMutableDictionary<NSString*, FlutterTextInputView*>* autofillContext;
 @property(nonatomic, assign) FlutterTextInputView* activeView;
+@property(nonatomic, retain) FlutterTextInputViewAccessibilityHider* inputHider;
```
nit: s/retain/strong for forward compatibility with ARC

gaaclarke reviewed (approved):
Objc code looks good to me.  I'm not keen on the approach but I believe you that this is the best way forward.  Thanks for the investigation =)

yasargil commented:
Can i request that this change get cherry picked into the 1.26 release?
